File: python/routes/routesPut.py
import logging
import os
import base64
import cgi
import http.cookies
import jwt
import json
import bcrypt
import urllib.parse
from http.server import BaseHTTPRequestHandler
from db.dbOperations import select_user_by_name, change_username, update_post_image, update_post_video, edit_comment_by_author, update_avatar
from middleware.jwt import verify_jwt

# Configuração do logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class routesPut(BaseHTTPRequestHandler):

    def do_PUT(self):
        routes = {
            '/change-username': self.handle_change_username,
            '/editpost': self.handle_editpost,
            '/edit-comment': self.handle_edit_comment,
            '/update-avatar': self.handle_update_avatar
        }

        if self.path in routes:
            routes[self.path]()
        else:
            self.handle_404()

    def handle_404(self):
        self.send_response(404)
        self.send_header('Content-type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps({"error": "Not Found"}).encode('utf-8'))

    def send_error_response(self, code, message):
        self.send_response(code)
        self.send_header('Content-type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps({"error": message}).encode('utf-8'))

    @verify_jwt
    def handle_change_username(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length).decode('utf-8')

            user_data = json.loads(post_data)
            new_name = user_data.get('username')

            user_name = self.decoded_token.get('name_user')
            user = select_user_by_name(user_name)

            if user:
                user_id = user['id_user']
                message = change_username(user_id, new_name)
                if message == "Username updated successfully.":
                    # Regenerar o token JWT com o novo nome de usuário
                    token = jwt.encode({'name_user': new_name}, os.getenv('JWT_SECRET'), algorithm='HS256')
                    
                    # Enviar o novo token JWT no cookie de resposta
                    cookie = http.cookies.SimpleCookie()
                    cookie['jwt_token'] = token
                    cookie['jwt_token']['path'] = '/'
                    
                    self.send_response(200)
                    self.send_header('Content-Type', 'application/json')
                    self.send_header('Set-Cookie', cookie.output(header=''))
                    self.end_headers()
                    self.wfile.write(json.dumps({'success': True, 'message': message}).encode('utf-8'))
                else:
                    self.send_response(400)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    self.wfile.write(json.dumps({'success': False, 'message': message}).encode('utf-8'))
            else:
                self.send_error_response(404, "User not found")
        except Exception as e:
            logger.error(f"Exception during change username: {e}")
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'success': False, 'message': str(e)}).encode('utf-8'))

    @verify_jwt
    def handle_editpost(self):
        try:
            content_type = self.headers.get('Content-Type')
            content_length = int(self.headers['Content-Length'])

            if 'multipart/form-data' in content_type:
                fields = cgi.FieldStorage(fp=self.rfile, headers=self.headers, environ={'REQUEST_METHOD': 'POST', 'CONTENT_TYPE': content_type})
                post_content = {key: fields[key].value for key in fields if fields[key].value}

                post_id = post_content.get('post_id')
                title = post_content.get('title')
                content = post_content.get('content')

                # Atualiza o post no banco de dados

                if 'file' in fields:
                    file_field = fields['file']
                    if self.is_valid_file_field(file_field):
                        file_data, file_name, file_size, file_type = self.read_file_field(file_field)
                        is_image, file_base64 = self.handle_file_upload(file_data, file_size, file_type)

                        if is_image:
                            update_post_image(post_id, title, content, file_base64)
                        else:
                            update_post_video(post_id, title, content, file_base64)

                    else:
                        self.send_error_response(400, "File field is empty or invalid")

                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'message': 'Post atualizado com sucesso'}).encode('utf-8'))

            else:
                self.send_error_response(400, "Invalid content type")
        except Exception as e:
            logger.error(f"Exception during update post: {e}")
            self.send_error_response(500, f"Server Error: {e}")
            
    @verify_jwt
    def handle_edit_comment(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length).decode('utf-8')

            comment_data = json.loads(post_data)
            
            comment = comment_data.get('comment')
            user_id = comment_data.get('user_id')
            post_id = int(comment_data.get('post_id'))
            id_comment = comment_data.get('id_comment')

            message = edit_comment_by_author(comment, user_id, post_id, id_comment)

            if message == "Comment updated successfully.":
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'message': message}).encode('utf-8'))
            else:
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': False, 'message': message}).encode('utf-8'))
        except Exception as e:
            logger.error(f"Exception during edit comment: {e}")
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'success': False, 'message': str(e)}).encode('utf-8'))
    @verify_jwt
    def handle_update_avatar(self):
        try:
            content_type = self.headers.get('Content-Type')
            content_length = int(self.headers['Content-Length'])
            user_name = self.decoded_token.get('name_user')
            user = select_user_by_name(user_name)
            user_id = user['id_user']

            if 'multipart/form-data' in content_type:
                fields = cgi.FieldStorage(fp=self.rfile, headers=self.headers, environ={'REQUEST_METHOD': 'PUT', 'CONTENT_TYPE': content_type})
                if 'avatar' in fields:
                    file_field = fields['avatar']
                    if self.is_valid_file_field(file_field):
                        file_data, file_name, file_size, file_type = self.read_file_field(file_field)
                        is_image, file_base64 = self.handle_file_upload(file_data, file_size, file_type)

                        if is_image:
                            update_avatar(user_id, file_base64)
                            self.send_response(200)
                            self.send_header('Content-Type', 'application/json')
                            self.end_headers()
                            self.wfile.write(json.dumps({'success': True, 'message': 'Avatar atualizado com sucesso'}).encode('utf-8'))
                        else:
                            self.send_error_response(400, "Invalid file type. Please upload an image file.")
                    else:
                        self.send_error_response(400, "File field is empty or invalid")
                else:
                    self.send_error_response(400, "File field is empty or invalid")
            else:
                self.send_error_response(400, "Invalid content type")
        except Exception as e:
            logger.error(f"Exception during update avatar: {e}")
            self.send_error_response(500, f"Server Error: {e}")

File: nodejs/Routes/routesPut.js
const fs = require("fs");
const path = require("path");
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");
const db = require("../db/dbOperations");
const authenticateToken = require("../middleware/jwt");
const formidable = require("formidable");
require("dotenv").config();

function setupPutRoutes(req, res) {
  const parsedUrl = require("url").parse(req.url, true);
  const pathName = parsedUrl.pathname;

  const putRoutes = {
    "/change-username": handleChangeUsername,
    "/editpost": handleEditPost,
    "/edit-comment": handleEditComment,
    "/update-avatar": handleUpdateAvatar,
  };

  if (req.method === "PUT" && putRoutes[pathName]) {
    putRoutes[pathName](req, res);
  } else if (req.method === "PUT") {
    res.writeHead(404, { "Content-Type": "text/plain" });
    res.end("Not Found");
  } else {
    res.writeHead(405, { "Content-Type": "text/plain" });
    res.end("Method Not Allowed");
  }
}

async function handleChangeUsername(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({ success: false, message: "Erro no servidor." })
        );
        return;
      }

      const newName = fields.username;
      const userName = req.user.name_user;

      console.log(newName, userName);

      try {
        const user = await db.selectUserByName(userName);
        if (!user || user.length === 0) {
          res.writeHead(404, { "Content-Type": "application/json" });
          res.end(
            JSON.stringify({
              success: false,
              message: "Usuário não encontrado",
            })
          );
          return;
        }

        const userId = user[0].id_user;
        const message = await db.changeUsername(userId, newName);

        if (message === "Username updated successfully.") {
          const token = jwt.sign(
            { name_user: newName },
            process.env.JWT_SECRET,
            { algorithm: "HS256" }
          );
          res.setHeader("Set-Cookie", `jwt_token=${token}; Path=/; HttpOnly`);
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: true, message: message }));
        } else {
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: false, message: message }));
        }
      } catch (err) {
        console.error(`Erro ao mudar o nome: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            success: false,
            message: `Erro no servidor: ${err.message}`,
          })
        );
      }
    });
  });
}


async function handleEditPost(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm({
      allowEmptyFiles: true,
      minFileSize: 0,
    });

    form.parse(req, async (err, fields, files) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ success: false, message: "Erro no servidor." }));
        return;
      }

      const postId = fields.post_id ? fields.post_id[0] : null;
      const title = fields.title ? fields.title[0] : null;
      const content = fields.content ? fields.content[0] : null;

      try {
        if (isValidFileField(files.file)) {
          const file = files.file;
          const { isImage, fileBase64 } = handleFileUpload(file);

          if (isImage) {
            await db.updatePostImage(postId, title, content, fileBase64);
          } else {
            await db.updatePostVideo(postId, title, content, fileBase64);
          }

          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: true, message: "Post atualizado com sucesso." }));
        } else {
          await db.updatePost(postId, title, content);
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: true, message: "Post atualizado com sucesso." }));
        }
      } catch (err) {
        console.error(`Erro ao atualizar o post: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ success: false, message: `Erro no servidor: ${err.message}` }));
      }
    });
  });
}


async function handleEditComment(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({ success: false, message: "Erro no servidor." })
        );
        return;
      }

      const comment = fields.comment;
      const userId = fields.user_id;
      const postId = parseInt(fields.post_id, 10);
      const idComment = fields.id_comment;

      try {
        const message = await db.editCommentByAuthor(
          comment,
          userId,
          postId,
          idComment
        );

        if (message === "Comment updated successfully.") {
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: true, message: message }));
        } else {
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ success: false, message: message }));
        }
      } catch (err) {
        console.error(`Erro ao editar o comentário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            success: false,
            message: `Erro no servidor: ${err.message}`,
          })
        );
      }
    });
  });
}

async function handleUpdateAvatar(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields, files) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({ success: false, message: "Erro no servidor." })
        );
        return;
      }

      const userName = req.user.name_user;
      const user = await db.selectUserByName(userName);
      const userId = user[0].id_user;
      const avatar = files.avatar;

      if (avatar) {
        const filePath = avatar[0].filepath;
        const fileType = avatar[0].mimetype;

        if (
          fileType.startsWith("image/jpeg") ||
          fileType.startsWith("image/png") ||
          fileType.startsWith("image/gif")
        ) {
          const fileData = fs.readFileSync(filePath);
          const fileBase64 = Buffer.from(fileData).toString("base64");

          try {
            await db.updateAvatar(userId, fileBase64);
            res.writeHead(200, { "Content-Type": "application/json" });
            res.end(
              JSON.stringify({
                success: true,
                message: "Avatar atualizado com sucesso",
              })
            );
          } catch (err) {
            res.writeHead(500, { "Content-Type": "application/json" });
            res.end(
              JSON.stringify({
                success: false,
                message: `Erro no servidor: ${err.message}`,
              })
            );
          }
        } else {
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(
            JSON.stringify({
              success: false,
              message: "Invalid file type. Please upload an image file.",
            })
          );
        }
      } else {
        res.writeHead(400, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            success: false,
            message: "File field is empty or invalid",
          })
        );
      }
    });
  });
}


function isValidFileField(file) {
  return file && file[0].size > 0 && file[0].mimetype;
}

function handleFileUpload(file) {
  const imageTypes = ["image/gif", "image/jpeg", "image/png"];
  const videoTypes = ["video/mp4"];
  const maxFileSize = 10 * 1024 * 1024;

  const fileData = fs.readFileSync(file[0].filepath);
  const fileSize = file[0].size;
  const fileType = file[0].mimetype;

  if (imageTypes.includes(fileType) && fileSize <= maxFileSize) {
    return { isImage: true, fileBase64: fileData.toString("base64") };
  } else if (videoTypes.includes(fileType) && fileSize <= maxFileSize) {
    return { isImage: false, fileBase64: fileData.toString("base64") };
  } else if (fileSize === 0) {
    throw new Error("Arquivo está vazio");
  } else {
    throw new Error("Tipo de arquivo ou tamanho inválido");
  }
}

module.exports = setupPutRoutes;
